onnxruntime/test/global_thread_pools/test_inference.cc

```diff
@@ -0,0 +1,222 @@
+class CApiTestGlobalThreadPoolsWithProvider : public testing::Test, public ::testing::WithParamInterface<int> {
+};
+
+template <typename OutT>
+static void RunSession(OrtAllocator* allocator, Ort::Session& session_object,
+                       const std::vector<Input>& inputs,
+                       const char* output_name,
+                       const std::vector<int64_t>& dims_y,
+                       const std::vector<OutT>& values_y,
+                       Ort::Value* output_tensor) {
+  std::vector<Ort::Value> ort_inputs;
+  std::vector<const char*> input_names;
+  for (size_t i = 0; i < inputs.size(); i++) {
+    input_names.emplace_back(inputs[i].name);
+    ort_inputs.emplace_back(Ort::Value::CreateTensor<float>(allocator->Info(allocator), const_cast<float*>(inputs[i].values.data()), inputs[i].values.size(), inputs[i].dims.data(), inputs[i].dims.size()));
+  }
+
+  std::vector<Ort::Value> ort_outputs;
+  if (output_tensor)
+    session_object.Run(Ort::RunOptions{nullptr}, input_names.data(), ort_inputs.data(), ort_inputs.size(), &output_name, output_tensor, 1);
+  else {
+    ort_outputs = session_object.Run(Ort::RunOptions{nullptr}, input_names.data(), ort_inputs.data(), ort_inputs.size(), &output_name, 1);
+    ASSERT_EQ(ort_outputs.size(), 1u);
+    output_tensor = &ort_outputs[0];
+  }
+
+  auto type_info = output_tensor->GetTensorTypeAndShapeInfo();
+  ASSERT_EQ(type_info.GetShape(), dims_y);
+  //size_t total_len = type_info.GetElementCount();
+  ASSERT_EQ(values_y.size(), static_cast<size_t>(5));
+
+  OutT* f = output_tensor->GetTensorMutableData<OutT>();
+  for (size_t i = 0; i != static_cast<size_t>(5); ++i) {
+    ASSERT_TRUE(abs(values_y[i] - f[i]) < 1e-6);
+  }
+}
+
+template <typename T, typename OutT>
+static Ort::Session GetSessionObj(Ort::Env& env, T model_uri, int provider_type) {
+  Ort::SessionOptions session_options;
+  session_options.DisablePerSessionThreads();
+
+  if (provider_type == 1) {
+#ifdef USE_CUDA
+    Ort::ThrowOnError(OrtSessionOptionsAppendExecutionProvider_CUDA(session_options, 0));
+    std::cout << "Running simple inference with cuda provider" << std::endl;
+#else
+    return Ort::Session(nullptr);
+#endif
+  } else if (provider_type == 2) {
+#ifdef USE_DNNL
+    Ort::ThrowOnError(OrtSessionOptionsAppendExecutionProvider_Dnnl(session_options, 1));
+    std::cout << "Running simple inference with dnnl provider" << std::endl;
+#else
+    return Ort::Session(nullptr);
+#endif
+  } else if (provider_type == 3) {
+#ifdef USE_NUPHAR
+    Ort::ThrowOnError(OrtSessionOptionsAppendExecutionProvider_Nuphar(session_options, /*allow_unaligned_buffers*/ 1, ""));
+    std::cout << "Running simple inference with nuphar provider" << std::endl;
+#else
+    return Ort::Session(nullptr);
+#endif
+  } else {
+    std::cout << "Running simple inference with default provider" << std::endl;
+  }
+
+  // if session creation passes, model loads fine
+  return Ort::Session(env, model_uri, session_options);
+}
```

Comment:
I was expecting some infrastructure similar to InferenceSessionGetGraphWrapper (although it should get graph_ via a protected accessor not directly, but that's a different matter) whereby you created a derived InferenceSession class that allowed you to check it was created the way you expected. Is there something in these tests that validates if the InferenceSession is actually using the global threadpool when expected or do they just test it doesn't break?

Reply:
When DisablePerSessionThreads is called and the right env constructor is used to create the global threadpools, env stores a flag called create_global_thread_pools_ = true obtained from EnvCreatedWithGlobalThreadPools(). This flag is checked for truth in ConstructoCommon to ensure the right threadpool ptrs are getting assigned. Unfortunately null checks for the threadpool ptrs don't work since they *can* be null. See 
    ORT_ENFORCE(session_env.EnvCreatedWithGlobalThreadPools() == true,
                "When the session is not configured to use per session"
                " threadpools, the env must be created with the the CreateEnvWithGlobalThreadPools API.");

Reply:
If I was to accidentally change ConstructorCommon to not do the right thing, would any tests here fail? Not required but would be nice if there was a sanity check done here. In this case it would be hard to break it, but in general I'm wary when unit tests don't explicitly test what they claim to (e.g. I have seen tests calling an endpoint to validate results that just checked there were results and no error, and not that the results were correct). 

You could potentially get the session state and check if the pointer for each threadpool matched the global ones and validate that against what was expected. Up to you.

---
In reply to: [393470632](https://github.com/microsoft/onnxruntime/pull/3177#discussion_r393470632) [](ancestors = 393470632)